buffer.go

```diff
@@ -11,142 +10,99 @@ import (
 	"github.com/pborman/uuid"
 )
 
-// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
-var PutRecordsLimit = 500
+var ErrRecordTooBig = errors.New("data byte size is over the limit")
 
-// RecordSizeLimit is the maximum allowed size per record.
-var RecordSizeLimit int = 1 * 1024 * 1024 // 1MB
+var (
+	// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
+	PutRecordsLimit int = 500
 
-// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
-var PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+	// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
+	PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
 
-type recordBuffer struct {
-	client   *kinesis.Kinesis
-	pKeyTmpl *template.Template
-	input    *kinesis.PutRecordsInput
+	// RecordSizeLimit is the maximum allowed size per record.
+	RecordSizeLimit int = 1 * 1024 * 1024 // 1MB
+)
+
+type buffer struct {
 	count    int
 	byteSize int
-	mutex    sync.Mutex
+	pKeyTmpl *template.Template
+	input    kinesis.PutRecordsInput
+	limits   map[string]int
 }
 
-func newRecordBuffer(client *kinesis.Kinesis, streamName string) (*recordBuffer, error) {
-	pKeyTmpl, err := compileTmpl("KINESIS_PARTITION_KEY_TEMPLATE")
-	if err != nil {
-		return nil, err
+func newBuffer(tmpl *template.Template, sn string) *buffer {
+	limits := map[string]int{
+		"PutRecordsLimit":     PutRecordsLimit,
+		"PutRecordsSizeLimit": PutRecordsSizeLimit,
+		"RecordSizeLimit":     RecordSizeLimit,
 	}
```

Comment:
I'd go with regular struct fields or perhaps a sub struct instead of a map. No need for a map since the keys never change, plus a little more error prone and less memory efficient.